### PR TITLE
Join some overloads of string methods to union parameters

### LINKED ISSUE #5766: Some signatures should use union instead of overloads
I think, this code should be correct code by fixes of lib.d.ts or type system.

``` ts
var separator: string|RegExp;
'str'.split(separator); // type error
```
## Manual: Replace to union types

Resolve by type definition.

from:

``` ts
// lib.d.ts
    split(separator: string, limit?: number): string[];
    split(separator: RegExp, limit?: number): string[];
```

to:

``` ts
// lib.d.ts
    split(separator: string|RegExp, limit?: number): string[];
```
## Automatic: Generate union types

Resolve by auto generated merged signature from overloads.

from:

``` ts
// lib.d.ts
    split(separator: string, limit?: number): string[];
    split(separator: RegExp, limit?: number): string[];
```

Generate the merged signature internally, implicitly.

to:

``` ts
// in memory
    split(separator: string, limit?: number): string[];
    split(separator: RegExp, limit?: number): string[];
    split(separator: string|RegExp, limit?: number): string[]; // generated
```

### PR DESCRIPTION
**Pull request by avonwyss**
Fixes the string methods which accept either a string or RegExp as pattern. Also, some help texts were changed to represent the actual behavior as per ECMA spec (e.g. when it is a literal pattern and when a regular expression pattern).

Fixes #5766

### CODE REVIEW

File: lib/lib.d.ts [hunk 1]
```diff
@@ -349,55 +349,29 @@ interface String {
...
     /**
-      * Matches a string with a regular expression, and returns an array containing the results of that search.
-      * @param regexp A regular expression object that contains the regular expression pattern and applicable flags.
-      */
-    match(regexp: RegExp): RegExpMatchArray | null;
-
-    /**
-      * Replaces text in a string, using a regular expression or search string.
-      * @param searchValue A string that represents the regular expression.
+      * Replaces text in a string, using a regular expression or search string. Only regular expression with global flag replaces all occurrences.
```
Comment (DanielRosenwasser) [inline lib/lib.d.ts]: Can you reword this to "When given a regular expression with its global flag set, this method replaces all occurrences. Otherwise, only the first match is replaced."

**Comment by falsandtru** [comment]
@avonwyss @DanielRosenwasser You need to edit declarations of the src directory. See https://github.com/Microsoft/TypeScript/blob/master/lib/README.md

https://github.com/Microsoft/TypeScript/blob/master/src/lib/es5.d.ts

**Comment by falsandtru** [comment]
any updates?

**Comment by RyanCavanaugh** [comment]
@falsandtru can you fix the merge conflict please? Sorry about that

Ping @DanielRosenwasser for review

**Comment by falsandtru** [comment]
done in #14639

**Comment by mhegazy** [comment]
Fixed by https://github.com/Microsoft/TypeScript/pull/14639
